File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2020/Y20D11.kt
package org.gristle.pdxpuzzles.advent.y2020

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.iteration.stabilized
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.objects.Grid
import org.gristle.pdxpuzzles.utilities.objects.mapToGridIndexed
import org.gristle.pdxpuzzles.utilities.objects.toGrid

class Y20D11(input: String) : Day {
    /**
     * Represents the states of each potential seat.
     */
    enum class Seat {
        OCCUPIED, UNOCCUPIED, EMPTY_SPACE;

        // Factory method
        companion object {
            fun of(char: Char): Seat {
                return when (char) {
                    '#' -> OCCUPIED
                    'L' -> UNOCCUPIED
                    '.' -> EMPTY_SPACE
                    else -> throw IllegalArgumentException("Input contains unrecognized char: $char")
                }
            }
        }

        // Convenience functions for semantic clarity.
        fun isOccupied() = this == OCCUPIED
        fun isEmptySpace() = this == EMPTY_SPACE
    }

    // Parse initial seating layout
    private val layout = input.toGrid(Seat::of)

    /**
     * Solves both parts using different tolerance levels and different algorithms to determine seat occupation
     * depending on the requirements for the two parts.
     */
    private fun solve(tolerance: Int, getNeighbors: (Grid<Seat>, Int) -> Int): Int {
        // Create a sequence that starts with the original layout and provides successive seat layouts, one per turn
        // The sequence uses the part-specific tolerance parameter and getNeighbor function to generate new states.
        val newStateSequence = generateSequence(layout) {
            it.mapToGridIndexed { index, seat -> // create new Grid
                if (seat.isEmptySpace()) { // empty floor space stays empty
                    Seat.EMPTY_SPACE
                } else {
                    val isOccupied = seat.isOccupied() // looks at whether the current seat is occupied
                    val neighbors = getNeighbors(it, index) // finds number of occupied seats using part-specific algo
                    // uses tolerance parameter to determine whether the seat becomes occupied or unoccupied
                    if ((isOccupied && neighbors < tolerance) || (!isOccupied && neighbors == 0)) {
                        Seat.OCCUPIED
                    } else {
                        Seat.UNOCCUPIED
                    }
                }
            }
        }

        // take the sequence; pair them up with the next iteration; find the first time where the next iteration 
        // does not change (i.e., the pattern has stabilized); then count the number of occupied seats.
        return newStateSequence
            .stabilized()
            .count(Seat::isOccupied)
    }

    override fun part1(): Int {
        // pt1 algo for finding occupied neighbors looks at seats immediately adjacent and counts those occupied 
        val getNeighbors = { grid: Grid<Seat>, index: Int ->
            grid.getNeighbors(index, true).count { seat -> seat.isOccupied() }
        }
        return solve(4, getNeighbors)
    }

    override fun part2(): Int {
        // pt2 algo for finding occupied neighbors, looks in all directions and returns the number of directions
        // where an occupied seat is found before an unoccupied seat.
        val getNeighbors = { grid: Grid<Seat>, index: Int ->
            // get the starting coordinate
            val startingCoord = grid.coordOf(index)
            // get 8 coordinates representing the slopes to calculate movement along the 8 directions from the
            // starting coordinate
            
            Coord.ALL_ADJACENT // for each direction...
                .map { slope -> // ...create a sequence of seats extending from line of sight, looking to see if an 
                    // occupied seat is encountered before an unoccupied seat.
                    // If it sees an unoccupied seat, the sequence ends without yielding a value because there is no 
                    // possibility that there will be an occupied seat in that direction per the part 2 rules.
                    sequence {
                        var newCoord = startingCoord + slope // mutable internal state: the next seat in line of sight
                        // yields values until every seat in that direction has been examined, or finds unoccupied seat 
                        while (grid.validCoord(newCoord)) {
                            if (grid[newCoord] == Seat.UNOCCUPIED) {
                                break
                            } else {
                                yield(grid[newCoord])
                            }
                            newCoord += slope // move to next seat along slope
                        }
                    }
                }
                // count the # directions where traversing the sequence finds an occupied seat.
                .count { lookSequence -> lookSequence.any(Seat::isOccupied) }
        }
        return solve(5, getNeighbors)
    }
}

fun main() = Day.runDay(Y20D11::class)

//    Class creation: 9ms
//    Part 1: 2243 (276ms)
//    Part 2: 2027 (371ms)
//    Total time: 658ms

File: rust/advent/src/bin/2020/y20d11.rs
use std::iter::successors;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::structs::{
    coord::Coord,
    grid::{Grid, Grid2, GridIterator},
    stopwatch::{ReportDuration, Stopwatch},
};

type Input = Grid2<Seat>;
type Output = usize;
type Pos = Coord<i32, 2>;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(20, 11).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum Seat {
    Occupied,
    Unoccupied,
    EmptySpace,
}
impl Seat {
    fn is_occupied(&self) -> bool {
        matches!(&self, Self::Occupied)
    }
    fn is_empty_place(&self) -> bool {
        matches!(&self, Self::EmptySpace)
    }
}

impl From<char> for Seat {
    fn from(value: char) -> Self {
        match value {
            '#' => Self::Occupied,
            'L' => Self::Unoccupied,
            '.' => Self::EmptySpace,
            c => {
                panic!("{c} is not a recognized character.");
            }
        }
    }
}

fn parse_input(input: &str) -> Input {
    Grid::new2d_map_str(input, Seat::from).unwrap()
}

fn solve<F>(layout: &Input, tolerance: Output, get_neighbors: F) -> Output
where
    F: Fn(&Input, Output) -> Output,
{
    // Create a sequence that starts with the original layout and provides successive seat layouts, one per turn
    // The sequence uses the part-specific tolerance parameter and getNeighbor function to generate new states.
    let width = layout.width();
    let new_states = successors(Some(layout.clone()), |layout| {
        let next_layout = layout
            .iter()
            .enumerate()
            .map(|(index, &seat)| {
                if seat.is_empty_place() {
                    Seat::EmptySpace
                } else {
                    let is_occupied = seat.is_occupied();
                    let neighbors = get_neighbors(&layout, index);
                    if (is_occupied && neighbors < tolerance) || (!is_occupied && neighbors == 0) {
                        Seat::Occupied
                    } else {
                        Seat::Unoccupied
                    }
                }
            })
            .try_collect_grid(width)
            .unwrap();
        Some(next_layout)
    });
    new_states
        .tuple_windows()
        .find(|(a, b)| a == b)
        .unwrap()
        .0
        .into_iter()
        .filter(Seat::is_occupied)
        .count()
}

fn part1(layout: &Input) -> Output {
    let get_neighbors = |layout: &Grid2<Seat>, index: Output| {
        layout
            .adjacent(index, true)
            .unwrap()
            .filter(|adj| adj.value.is_occupied())
            .count()
    };
    solve(layout, 4, get_neighbors)
}

fn part2(layout: &Input) -> Output {
    let width = layout.width();
    let get_neighbors = |layout: &Grid2<Seat>, index: Output| {
        let pos = Pos::new2d((index % width) as i32, (index / width) as i32);
        vec![
            Pos::new2d(-1, -1),
            Pos::new2d(0, -1),
            Pos::new2d(1, -1),
            Pos::new2d(-1, 0),
            Pos::new2d(1, 0),
            Pos::new2d(-1, 1),
            Pos::new2d(0, 1),
            Pos::new2d(1, 1),
        ]
        .into_iter()
        .filter(|&slope| {
            let mut new_pos = pos + slope;
            let mut keep = false;
            while let Some(seat) = layout.get(new_pos) {
                match seat {
                    Seat::Occupied => {
                        keep = true;
                        break;
                    }
                    Seat::Unoccupied => {
                        break;
                    }
                    Seat::EmptySpace => {
                        new_pos += slope;
                    }
                }
            }
            keep
        })
        .count()
    };
    solve(layout, 5, get_neighbors)
}

#[test]
fn default() {
    let input = get_input(20, 11).unwrap();
    let input = parse_input(&input);
    assert_eq!(2243, part1(&input));
    assert_eq!(2027, part2(&input));
}

// Input parsed (71μs)
// 1. 2243 (85ms)
// 2. 2027 (39ms)
// Total: 124ms
